python/pyiceberg/io/fsspec.py

```diff
@@ -0,0 +1,201 @@
+
+    def exists(self) -> bool:
+        """Checks whether the location exists"""
+        return self._fs.lexists(self.location)
+
+    def create(self, overwrite: bool = False):
+        """Create an output stream for reading the contents of the file
+
+        Args:
+            overwrite(bool): Whether to overwrite the file if it already exists
+
+        Returns:
+            OpenFile: An fsspec compliant file-like object
+
+        Raises:
+            FileExistsError: If the file already exists at the location and overwrite is set to False
+
+        Note:
+            If overwrite is set to False, a check is first performed to verify that the file does not exist.
+            This is not thread-safe and a possibility does exist that the file can be created by a concurrent
+            process after the existence check yet before the output stream is created. In such a case, the default
+            behavior will truncate the contents of the existing file when opening the output stream.
+        """
+        if not overwrite and self.exists():
+            raise FileExistsError(f"Cannot create file, file already exists: {self.location}")
+        return self._fs.open(self.location, "wb")
+
+    def to_input_file(self) -> FsspecInputFile:
+        """Returns a new FsspecInputFile for the location at `self.location`"""
+        return FsspecInputFile(location=self.location, fs=self._fs)
+
+
+class FsspecFileIO(FileIO):
+    """A FileIO implementation that uses fsspec"""
+
+    def __init__(self, properties: Properties):
+        self._scheme_to_fs = {}
+        self._scheme_to_fs.update(SCHEME_TO_FS)
+        super().__init__(properties=properties)
+
+    def new_input(self, location: str) -> FsspecInputFile:
+        """Get an FsspecInputFile instance to read bytes from the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecInputFile: An FsspecInputFile instance for the given location
+        """
+        uri = urlparse(location)
+        
+        if uri.scheme not in self._scheme_to_fs:
+            raise ValueError(f"No registered filesystem for scheme: {uri.scheme}")
+        fs = self._scheme_to_fs[uri.scheme](self.properties, **self._fs_properties())
+        return FsspecInputFile(location=location, fs=fs)
+
+    def new_output(self, location: str) -> FsspecOutputFile:
+        """Get an FsspecOutputFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): A URI or a path to a local file
+
+        Returns:
+            FsspecOutputFile: An FsspecOutputFile instance for the given location
+        """
+        uri = urlparse(location)
+        if uri.scheme not in self._scheme_to_fs:
+            raise ValueError(f"No registered filesystem for scheme: {uri.scheme}")
+        fs = self._scheme_to_fs[uri.scheme](self.properties, **self._fs_properties())
```

Comment:
I think these need to be cached. It looks like this is going to create a new `S3FileSystem` for every input or output file. We should build just one and reuse it every time we see the scheme.

Reply:
Ah right, I added `self._fs_cache: Dict[str, AbstractFileSystem] = {}` to the `FsspecFileIO` class to cache the filesystem based on the uri scheme and then I check that first before creating a new one.